src/librustc_mir/transform/const_prop.rs

```diff
@@ -334,7 +352,24 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
                 trace!("evaluated promoted {:?} to {:?}", promoted, res);
                 Some(res.into())
             },
-            _ => None,
+            Place::Base(
+                PlaceBase::Static(box Static {kind: StaticKind::Static(def_id), .. })
+            ) => {
+                // mutable statics can change value at any time so we can't const
+                // propagate their values
+                if self.tcx.is_mutable_static(def_id) {
```

Comment:
Does this also need to take internal mutability into account?

Reply:
It's even worse, the moment you encounter a union or raw pointer you need to give up. Is this change needed for this PR or could it be done in an extra PR?

Reply:
> could it be done in an extra PR?

Possibly? I had some trouble getting a `Len` operand to appear in the MIR and this was necessary to get the current test case I came up with to pass. The obvious test case

```rust
let x= &[1, 2, 3];
x[1]
```

[doesn't seem to generate a `Len` operand at all](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7179047e946f37426f027ef3c7a710de).

I'll look into this tonight but this might not be necessary to get this test case to work:

```rust
const TEST: &[u32] = &[1, 2, 3];

fn test() -> u32 {
    TEST[1]
}
```

Reply:
You get one if you use `let x: &[u8] = &[1, 2, 3];`. Arrays don't generate `Len` `Rvalue`s

Reply:
Oh wow. I didn't realize `&[1, 2, 3]` got inferred as a `&[{integer}; 3]` instead of a slice. That makes sense why my original test case didn't work then. A lot of these changes are probably not necessary then.

Reply:
So unfortunately this test case doesn't work because `x` is a local and we currently only const propagate temporaries: 

https://github.com/rust-lang/rust/blob/02f5786a324c40b2d8b2d0df98456e48fb45d30c/src/librustc_mir/transform/const_prop.rs#L576-L580

Reply:
hmm... The original test wasn't `((&[1, 2, 3]) as &[u32])[1]`, because I didn't what the MIR after the `&[1, 2, 3]` to know about the actual array size. But that's not really useful I guess. Feel free to just change it to one big expression and scrap the rest of the test.

Reply:
Sorry, my previous comment was kind of vague. So far, I haven't been able to find a test case that contains a `Len` operand that we can const propagate without the changes to handle statics. Here's what I've tried:

```rust
let x: &[u32] = &[1, 2, 3];
x[1];
```

This style test doesn't work at all because `x` is a local and we don't const propagate locals. 

```rust
(&[1, 2, 3] as &[u32])[1];
```

This doesn't compile because it's a non-primitive cast.

```rust
static X: &[u32] = &[u32];

X[1];
```

This *could* maybe work but would require additional changes like those included in this PR to handle statics.

```rust
const X: &[u32] = &[1, 2, 3];

X[1];
```

This I believe will work once we have support in `replace_with_const()` for `Operand::Indirect`.

So, as it stands currently, I don't think there's anyway to write a test which will exercise this part of the const propagator.

Reply:
The `let` binding case we can't do until we get some sort of dataflow analysis going

The static case scares me. I'm not sure we should be propagating statics, like ever. A static is something with a name. Reading from it seems like something we shouldn't optimize out, not sure how aggressive we can or want to be there. It's probably alright if the value being read is not from a static mut and the type has no interior mutability, but if we read beyond that (like from things with `UnsafeCell` or behind raw pointers) we have no guarantees. The problem is, once we read from a static, the propagated value has no notion of "I'm from a static" anymore. So we don't know anymore that we can't just read further values. Most likely not a problem for now, as propagating raw pointer derefs is its very own crazy topic that probably needs UGC discussions. But for now I'd rather not open this box.

The other two cases seem reasonable to me, pick either I guess? Though you may end up needing `Indirect` support for the cast case, too, so just go with the constant, which requires the least changes.

Reply:
> The static case scares me. I'm not sure we should be propagating statics, like ever.

Yeah, I agree. There's so many edge cases here that it would be very complicated to ensure this was done correctly.

> The let binding case we can't do until we get some sort of dataflow analysis going

Do we need full dataflow analysis or is there a simpler thing we could do now and eventually replace with dataflow? For example, I have a change I've been playing with that allows propagating locals which were themselves propagated into. This only is only at the in the `ConstPropagator` and not the `CanConstProp` pass so it doesn't affect error reporting. (See https://github.com/rust-lang/rust/commit/e699063288467f9169fcb7c61ddd80d36b5b0474 for my code)

Reply:
hmm... we'd also need to make sure to propagate neither *from* nor *into* constants that have more than one mutable usage. You don't want any const prop for `let x = if y { 5 } else { 6 };` or for `let mut x = 42; if y { x = 9; }`